profile-widget/profilewidget2.cpp

```diff
@@ -2044,70 +2045,75 @@ void ProfileWidget2::clearPictures()
+	std::sort(pictures.begin(), pictures.end());
 
 	double x, y, lastX = -1.0, lastY = -1.0;
-	for (int i = m->rowDDStart; i < m->rowDDEnd; i++) {
-		int picItemNr = i - m->rowDDStart;
-		int offsetSeconds = m->index(i, 1).data(Qt::UserRole).value<int>();
-		// it's a correct picture, but doesn't have a timestamp: only show on the widget near the
-		// information area. A null pointer in the pictures array indicates that this picture is not
-		// shown.
-		if (!offsetSeconds) {
-			pictures[picItemNr].reset();
+	int size = Thumbnailer::defaultThumbnailSize();
+	for (PictureEntry &e: pictures) {
+		// Only show pictures whose offsets are withing the profile
+		if (e.offset.seconds <= 0 || e.offset.seconds >= current_dive->duration.seconds)
 			continue;
-		}
-		DivePictureItem *item = pictures[picItemNr].get();
-		if (!item) {
-			item = new DivePictureItem;
-			pictures[picItemNr].reset(item);
-			scene()->addItem(item);
-		}
-		item->setVisible(prefs.show_pictures_in_profile);
-		item->setPixmap(m->index(i, 0).data(Qt::UserRole).value<QPixmap>());
-		item->setFileUrl(m->index(i, 1).data().toString());
+		e.thumbnail.reset(new DivePictureItem);
+		scene()->addItem(e.thumbnail.get());
+		e.thumbnail->setVisible(prefs.show_pictures_in_profile);
+		QImage thumbnail = Thumbnailer::instance()->fetchThumbnail(e.filename).scaled(size, size, Qt::KeepAspectRatio);
+		e.thumbnail->setPixmap(QPixmap::fromImage(thumbnail));
+		e.thumbnail->setFileUrl(e.filename);
 		// let's put the picture at the correct time, but at a fixed "depth" on the profile
 		// not sure this is ideal, but it seems to look right.
-		x = timeAxis->posAtValue(offsetSeconds);
-		if (i == 0)
+		x = timeAxis->posAtValue(e.offset.seconds);
+		if (lastX < 0.0)
 			y = 10;
 		else if (fabs(x - lastX) < 3 && lastY <= (10 + 14 * 3))
 			y = lastY + 3;
 		else
 			y = 10;
 		lastX = x;
 		lastY = y;
-		item->setPos(x, y);
+		e.thumbnail->setPos(x, y);
 	}
 }
 
-void ProfileWidget2::removePictures(const QModelIndex &, int first, int last)
+// TODO: This does not check for the fact that the same image may be attributed
+// to different dives! Deleting the picture from one dive may therefore remove
+// it from the profile of a different dive.
+void ProfileWidget2::removePictures(const QVector<QString> &fileUrls)
 {
-	DivePictureModel *m = DivePictureModel::instance();
-	first = std::max(0, first - m->rowDDStart);
-	// Note that last points *to* the last item and not *past* the last item,
-	// therefore we add 1 to achieve conventional C++ semantics.
-	last = std::min((int)pictures.size(), last + 1 - m->rowDDStart);
-	if (first >= (int)pictures.size() || last <= first)
-		return;
-	pictures.erase(pictures.begin() + first, pictures.begin() + last);
+	// Delete in ranges instead of single entries to avoid unnecessary copying. Premature optimization?
+	for (auto it1 = pictures.begin(); it1 != pictures.end(); ++it1) {
+		if (std::find(fileUrls.begin(), fileUrls.end(), it1->filename) == fileUrls.end())
```

Comment:
how / why is this different from the find_if with lambda thingy above at line 2050? It seems to do the same thing? And if not, that too might be worth a comment

Reply:
that on compares struct members and needs a special / separate unction (lambda) for the comparison.
this one compares QStrings directly as QString already has a `==` operator required by `std::find`.

Reply:
Note that this find is inverted compared to the find above.

In the above function we got a filename and search for the picture with that filename.

Here we got a list of filenames and for every picture we search whether its filename is in the list.

This function is indeed complex. The reason is that it tries to avoid doing repeated copying of vector elements, by deleting the the images in batches. It will find batches of continuous images to be deleted an remove them at once.